### ISSUE TITLE
Crash when edges are not initialised with a weight attribute

### ISSUE THREAD

**Issue by 8W9aG** [eliorc/node2vec#91]
I am seeing the following crash when creating a `Node2Vec` object on a graph with no weights attribute on its edges:

```
Traceback (most recent call last):
  File "~/dev/pyiextract/graph.py", line 65, in <module>
    node2vec = Node2Vec(G, dimensions=64, walk_length=30, num_walks=200, workers=4)
  File "~/dev/pyiextract/venv/lib/python3.9/site-packages/node2vec/node2vec.py", line 73, in __init__
    self._precompute_probabilities()
  File "~/dev/pyiextract/venv/lib/python3.9/site-packages/node2vec/node2vec.py", line 125, in _precompute_probabilities
    ss_weight = weight * 1 / q
TypeError: unsupported operand type(s) for *: 'NoneType' and 'int'
```

If I initialise the edges with a weights attribute of 0.0, it seems to work. I can make a PR to assume 0.0 as a weight attribute if none exists however I want to understand what the intention is here, perhaps if this is expected an explicit error could be sent as an exception instead?

**Comment by eliorc** [comment]
Hi, thanks for that

Do you mean that some edges have a weight and some don't? Or all edges have weight and some weights are 0? 

It will be best if you could supply code to reproduce the problem

**Comment by 8W9aG** [comment]
I mean my edges have no weights, code is as follows:

```
print("Creating a graph...")
G = nx.MultiDiGraph()
files = []
for file in glob.glob("output/*.json"):
    files.append(file)
    with open(file) as json_handle:
        doc = json.load(json_handle)
        for triple in doc["triples"]:
            head_entity = triple["head_entity"]
            tail_entity = triple["tail_entity"]
            head_entity_id = head_entity["identifier"]
            tail_entity_id = tail_entity["identifier"]
            connection = triple["connection"].strip().translate(str.maketrans('', '', string.punctuation))
            if not connection or len(connection) < 2:
                continue
            connection_id = "|".join([connection, str(triple["negated"] if triple["negated"] is not None else False).lower()])
            if head_entity_id not in G:
                G.add_node(head_entity_id, name=head_entity["name"])
            if tail_entity_id not in G:
                G.add_node(tail_entity_id, name=tail_entity["name"])
            found_edge_id = None
            for edge_id in range(G.number_of_edges(head_entity_id, tail_entity_id)):
                data = G.edges[head_entity_id, tail_entity_id, edge_id]
                if data["connection"] == connection_id:
                    found_edge_id = edge_id
                    break
            if found_edge_id is None:
                G.add_edge(head_entity_id, tail_entity_id, connection=connection_id, is_quote=triple["is_quote"], temporal=triple["temporal"], location=triple["location"])
            else:
                for triple_key in triple:
                    if triple_key == "head_entity" or triple_key == "tail_entity" or triple_key == "connection":
                        continue
                    if triple_key in G.edges[head_entity_id, tail_entity_id, found_edge_id]:
                        if G.edges[head_entity_id, tail_entity_id, found_edge_id][triple_key] is None:
                            continue
                    if triple[triple_key] is None:
                        continue
                    G.edges[head_entity_id, tail_entity_id, found_edge_id][triple_key] = triple[triple_key]
```

**Comment by eliorc** [comment]
I cannot execute this code... I meant a standalone code example

I have used this library multiple times with graphs that have no weights and it works, also the code in the README works and it has no weight assigned to edges. So I'm not sure where the problem is. It looks from the error that you set the `weight` attribute to `None`. I suggest that if you have a `None` or `0` weight, just don't include this edge in the graph

**Comment by 8W9aG** [comment]
The code that I am looking at in node2vec (from the pypi installation) is as follows:

```python
try:
                        if self.graph[current_node][destination].get(self.weight_key):
                            weight = self.graph[current_node][destination].get(self.weight_key)
                        else: 
                            ## Example : AtlasView({0: {'type': 1, 'weight':0.1}})- when we have edge weight
                            edge = list(self.graph[current_node][destination])[-1]
                            weight = self.graph[current_node][destination][edge].get(self.weight_key)
                            
                    except:
                        weight = 1 
                    
                    if destination == source:  # Backwards probability
                        ss_weight = weight * 1 / p
                    elif destination in self.graph[source]:  # If the neighbor is connected to the source
                        ss_weight = weight
                    else:
                        ss_weight = weight * 1 / q
```

It seems like this code with its use of `get` would produce a None weight if no weight is apparent. Here is the vanilla code that reproduces this issue:

```python
print("Creating a graph...")
G = nx.MultiGraph()
G.add_node(1)
G.add_node(2)
G.add_edge(1, 2)
node2vec = Node2Vec(G, dimensions=64, walk_length=30, num_walks=200, workers=4)
sys.exit()
```

**Comment by eliorc** [comment]
Thanks for this code, I can see the problem
I will look over this during the weekend I hope. I can see this does not happen with you use a `nx.Graph` instead of `nx.MultiGraph` I wonder why...

For example this code works

```python3
print("Creating a graph...")
G = nx.Graph()
G.add_node(1)
G.add_node(2)
G.add_edge(1, 2)
node2vec = Node2Vec(G, dimensions=64, walk_length=30, num_walks=200, workers=4)
```

**Comment by eliorc** [comment]
What is the type of edges you need in `MultiGraph`? From this picture, do you need support for the red or blue ones?
![image](https://user-images.githubusercontent.com/17727283/181434552-8a4c0d36-f204-45d9-ab48-17971207d01a.png)

**Comment by 8W9aG** [comment]
Both would be preferrable, I have a graph where nodes can be connected to other nodes via a link which is defined mainly by the connection attribute. For example "YYZ" -> "located in" -> "Toronto" and "YYZ" -> airport of -> "Toronto".

**Comment by eliorc** [comment]
Fixed in 0.4.6 (out on PyPI)
